color/src/parse.rs

```diff
@@ -482,43 +485,70 @@ pub fn parse_color(s: &str) -> Result<DynamicColor, ParseError> {
                     Err(ParseError::UnknownColorIdentifier)
                 }
             }
-        }
-        // TODO: should we validate that the parser is at eof?
+        }?;
+        Ok((parser.ix, color))
     } else {
         Err(ParseError::UnknownColorSyntax)
     }
 }
 
-const fn get_4bit_hex_channels(hex_str: &str) -> Result<[u8; 8], ParseError> {
-    let mut four_bit_channels = match *hex_str.as_bytes() {
-        [r, g, b] => [r, r, g, g, b, b, b'f', b'f'],
-        [r, g, b, a] => [r, r, g, g, b, b, a, a],
-        [r0, r1, g0, g1, b0, b1] => [r0, r1, g0, g1, b0, b1, b'f', b'f'],
-        [r0, r1, g0, g1, b0, b1, a0, a1] => [r0, r1, g0, g1, b0, b1, a0, a1],
-        _ => return Err(ParseError::WrongNumberOfHexDigits),
-    };
+// Arguably this should be an implementation of FromStr.
+/// Parse a color string in CSS syntax into a color.
+///
+/// This parses the entire string; trailing characters cause an
+/// [`ExpectedEndOfString`](ParseError::ExpectedEndOfString) parse error. See also
+/// [`parse_color_with_remainder`].
+///
+/// # Errors
+///
+/// Tries to return a suitable error for any invalid string, but may be
+/// a little lax on some details.
+pub fn parse_color(s: &str) -> Result<DynamicColor, ParseError> {
+    let (ix, color) = parse_color_with_remainder(s)?;
+
+    if ix == s.len() {
+        Ok(color)
+    } else {
+        Err(ParseError::ExpectedEndOfString)
+    }
+}
+
+/// Parse 4-bit color channels from a hex-encoded string.
+///
+/// Returns the parsed channels and the byte offset to the remainder of the string (i.e., the
+/// number of hex characters parsed).
+const fn get_4bit_hex_channels(hex_str: &str) -> Result<(usize, [u8; 8]), ParseError> {
+    let mut hex = [0; 8];
 
-    // convert to hex in-place
-    // this is written without a for loop to satisfy `const`
     let mut i = 0;
-    while i < four_bit_channels.len() {
-        let ascii = four_bit_channels[i];
-        let as_hex = match hex_from_ascii_byte(ascii) {
-            Ok(hex) => hex,
-            Err(e) => return Err(e),
-        };
-        four_bit_channels[i] = as_hex;
-        i += 1;
+    while i < 8 && i < hex_str.len() {
+        if let Ok(h) = hex_from_ascii_byte(hex_str.as_bytes()[i]) {
+            hex[i] = h;
+            i += 1;
+        } else {
+            break;
+        }
     }
-    Ok(four_bit_channels)
+
+    let four_bit_channels = match i {
```

Comment:
Can this use slice destructuring matching? I.e.
```
match &hex[..i] {
    [r, g, b] => [r, r, g, g, b, b, 15, 15],
// ...
    _ if hex.len() == 8 => hex,
 _ => return Err(...)
}
```

Reply:
I don't think we can:

```
533 |     let four_bit_channels = match &hex[..i] {
    |                                       ^^^^^
    |
    = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants
```

Reply:
(This function doesn't _have_ to be const, it just was, and I've kept it that way.)

Reply:
:(

Reply:
Yeah, unfortunately this code is not as nice as it could be because of const. I do think that's nice to have though.